src/app/plugins/githubuploadadapter.js

```diff
@@ -103,8 +103,32 @@ export class Adapter {
 						// Step 2: the real upload takes place this time to Amazon S3 servers,
 						// using information returned from Step 1.
 
-						// The final URL of the file is already known, even before the upload. We save it here.
-						returnUrl = response.asset.href;
+						const assetUrl = typeof response.asset_upload_url === 'string' ? response.asset_upload_url : null;
+						const authenticityToken = typeof response.asset_upload_authenticity_token == 'string' ?
+							response.asset_upload_authenticity_token : null;
+
+						if ( !( assetUrl && authenticityToken ) ) {
+							return;
+						}
+
+						const form = new FormData();
+
+						form.append( 'authenticity_token', authenticityToken );
+
+						fetch( assetUrl, {
+							method: 'PUT',
+							body: form,
+							credentials: 'same-origin',
+							headers: {
+								Accept: 'application/json',
+								'X-Requested-With': 'XMLHttpRequest'
+							}
+						} )
+							.then( response => response.json() )
+							.then( response => {
+								// The final URL of the file is already known, even before the upload. We save it here.
+								returnUrl = response.href;
```

Comment:
This is where the promise ([constructed in this line](https://github.com/ckeditor/github-writer/blob/4db90dd71dc58757feac652d439e519c085b4617/src/app/plugins/githubuploadadapter.js#L102)) should resolve.

It should resolve with passing the `response.href` so there will be no need for the [globalish variable `returnUrl`](https://github.com/ckeditor/github-writer/blob/4db90dd71dc58757feac652d439e519c085b4617/src/app/plugins/githubuploadadapter.js#L56).